### PR TITLE
Work In Progress: Finally fix CI

### PR DESCRIPTION
**Pull request by arrowcircle**
This is work in progress of fixing CI in master branch.  Do not merge yet.

I separated releases of NPM and pypy packages and made release pipeline trigger only on tags in master branch.

Also, it looks like not all configuration of buildkite is in the repo.

Is it possible to move all buildkite steps and configuration to the repo, so we have everything in one place?

Also, I think we should bundle js files in python package before having something stable. Releasing separate versions of JS and python packages just makes a lot of confusion. And npm package release does not make any sense right now, because Yukon does not use this package in any way.

UPD1.

Step one is to separate release pipeline to separate pipeline: `.buildkite/pipeline.release.yml` and trigger it from main pipeline.yml only on master and tags. Done. Needs review.

Next step is to divide JS and Python pipelines and run inside docker containers. Frontend and backend does not have interconnections and do not require to be run in one container with both python and nodejs. JS step separated in `':nodejs: Run JavaScript checks'` label. Done. Needs review.

Then pipelines of buildkite should be configured in yml file. This does have downside, because trigger requires a slug, and slug is available only via web interface. This is huge problem, because requires to do something in web interface => requires to config something in another place. Travis CI does look better because these web-interface steps are not needed. Who can help with this? Or should we move to Travis CI?

### CODE REVIEW

**Comment by thirtytwobits** [comment]
Let me know if we have UAVCAN organization members that need access to our Buildkite organization.

A few of things to consider:

1. Be careful what you allow any CI system to do from a PR. I setup a separate pipeline with steps that were in the Buildkite UI instead of pulling from the source so that random third parties could not use our CI as a zombie actor with elevated permissions.

2. We don't have a dedicated AWS account for UAVCAN but if we want one I can set this up. I've been using my personal account and I was able to get AWS to sponsor our project so I'd have to move our account credits over if we do want to do this.

3. I started using Buildkite because it offers maximum flexibility for CI workers. It's the only service I'm aware of that lets you insert arbitrary (but authenticated) resources into their pipelines. We're planning on using this for libuavcan to provide on-target testing for multiple platforms (e.g. we can actually put a Pixhawk hawk into a Buildkite pipeline if we wanted to). For Yukon this type of automation would be interesting to allow test rigs involving a full IP stack and remoting between two systems. See [Nanaimo](https://nanaimo.readthedocs.io/en/latest/docs/guide.html) for details.

This all said, if you find the complexity of Buildkite stifling (and, as CI services go, it is in the complex/powerful quadrant of the matrix) and want to switch to something else I'm not going to object. I am not the maintainer of Yukon. Only a hopeful customer and fan.

Thanks for contributing  everyone!

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+  - label: ':python: Run Pyhton checks'
+    command: tox -p auto -e "{py37,py38}-{test,doctest},pylint,mypy,report"
+
+  - wait
+
+  - label: "🚀Release"
+    trigger: "yukon-release"
+    branches: master
+    if: build.tag != null
```
Comment (thirtytwobits) [inline .buildkite/pipeline.yml:27]: nice. The problem is a PR can change the value of "branches:" to trigger a release. This is a security hole we cannot allow.

**Comment by arrowcircle** [comment]
@thirtytwobits I will fix security concern and will remove it from PR. Is it possible to share access to buildkite account so I can configure pipeline for both PR and release?

If it's possible to separate secrets between commit and release pipelines, we can easily store commit pipeline in the repo and store release pipeline in buildkite (I dont like this, but it it's the best way for now).

BTW, is it possible to add some restrictions on PR commits, so it's not possible to run releases from PR?

File: .buildkite/pipeline.yml [hunk 2]
```diff
@@ -18,10 +11,3 @@ steps:
 
   - label: ':python: Run Pyhton checks'
     command: tox -p auto -e "{py37,py38}-{test,doctest},pylint,mypy,report"
-
-  - wait
-
-  - label: "🚀Release"
```
Comment (thirtytwobits) [inline .buildkite/pipeline.yml]: Why did you remove this? Now that you have access to Buildkite you can ensure that this isn't used in the PR pipeline so it should be okay now.

**Comment by arrowcircle** [comment]
I made a quick look through the docs and interface of the buildkite and found, that interface differs from documentation. For now I will make pipeline templates files, that can be copied to the web interface of the buildkite.

I think main pipeline should be run on every event including master branch commits. Release pipeline should not run automatically, but will be triggered via main pipeline on tags and success of previous steps.

@thirtytwobits Does this make sense?

File: .buildkite/pipeline.yml.template [hunk 3]
```diff
@@ -0,0 +1,50 @@
+steps:
+  - label: ':nodejs: Run JavaScript checks'
+    plugins:
+      - docker#v3.5.0:
+          image: "node:12"
```
Comment (thirtytwobits) [inline .buildkite/pipeline.yml.template:5]: Who controls this image? Can we use the nodethon container instead? I'd like to control our containers so we don't have to chase build breaks because some other project released a buggy container.

File: .buildkite/release.sh [hunk 4]
```diff
@@ -33,6 +33,6 @@ export YUKON_FRONTEND_MAJOR_MINOR_VERSION=$(echo $YUKON_FRONTEND_FULL_VERSION |
 tox -p auto
 tox -e sonar-release | grep -v "sonar.login"
 # Upload yukon_backend to PyPi
-tox -e pypi-upload | grep -v "twine upload"
+# tox -e pypi-upload | grep -v "twine upload"
```
Comment (thirtytwobits) [inline .buildkite/release.sh:36]: can we just delete the commented out lines now? They'll just cause confusion.

File: .buildkite/pipeline.yml.template [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+          image: "uavcan/nodethon:node13-py37-py38"
+          propagate-environment: true
+    agents:
+      queue: 'default'
+    timeout_in_minutes: 15
+
+  - wait
+
+  - label: "🚀Release"
+    trigger: "yukon-release"
```
Comment (thirtytwobits) [inline .buildkite/pipeline.yml.template:48]: nice!

File: .buildkite/pipeline.yml.template (hunk 3 above)
Comment (arrowcircle) [reply .buildkite/pipeline.yml.template:5]: It does nod have namespace, it's official docker image of nodejs. I dont like the idea of using nodethon, because python is not needed for pure javascript part. And nodethon uses node v13, that is not LTS version. Also, nodethon does not have yarn inside.

File: .buildkite/release.sh (hunk 4 above)
Comment (thirtytwobits) [reply .buildkite/release.sh:36]: If it's not used then remove it.

Comment (arrowcircle) [reply .buildkite/release.sh]: done

File: .buildkite/pipeline.yml.template (hunk 3 above)
Comment (arrowcircle) [reply .buildkite/pipeline.yml.template:5]: What we decide here?

**Comment by arrowcircle** [comment]
@thirtytwobits It looks like we have one question about docker image. But main question is are we ready to update pipelines in buildkite web interface?

I am ready to change pr pipeline to content of `pipeline.yml.template`. Is it ok?

File: .buildkite/pipeline.yml.template (hunk 3 above)
Comment (pavel-kirienko) [reply .buildkite/pipeline.yml.template:5]: As discussed with Scott, we can move forward with a greater autonomy so if you think we should use the official image then that's that.

**Comment by pavel-kirienko** [comment]
> I am ready to change pr pipeline to content of pipeline.yml.template. Is it ok?

Please go ahead but mind the security as advised by Scott.

**Comment by arrowcircle** [comment]
I uploaded new yml to buildkite for pull requests. Will look into what will break.

**Comment by arrowcircle** [comment]
So what do we do?

**Comment by pavel-kirienko** [comment]
We accept this and we fix the tests afterwards.
